docs/internal/DistributedArchitectureGuide.md

```diff
@@ -269,6 +286,22 @@ of shards, and an incentive to distribute shards within the same index across di
 `NodeAllocationStatsAndWeightsCalculator` classes for more details on the weight calculations that support the `DesiredBalanceComputer`
 decisions.
 
+### Inter-Node Communicaton
+
+The elected master node creates a shard allocation plan with the `DesiredBalanceShardsAllocator` and then selects incremental shard
+movements towards the target allocation plan with the `DesiredBalanceReconciler`. The results of the `DesiredBalanceReconciler` is an
+updated `RoutingTable`. The `RoutingTable` is part of the cluster state, so the master node updates the cluster state with the new
+(incremental) desired shard allocation information. The updated cluster state is then published to the data nodes. Each data node will
+observe any change in shard allocation related to that node and take action to achieve the new shard allocation by initiating creation of a
+new empty shard, starting recovery (copying) of an existing shard from another data node, or remove a shard. When the data node finishes
```

Comment:
grammar nit:

```suggestion
new empty shard, starting recovery (copying) of an existing shard from another data node, or removing a shard. When the data node finishes
```

but also I feel we should expand a bit on what "removing a shard" means in this context? It means actively shutting down the corresponding running (or recovering) `IndexShard` instance, releasing all its resources.

Again, this feels like detail that should be in a Javadoc somewhere, maybe `org.elasticsearch.cluster.routing.GlobalRoutingTable`, and linked to from `RoutingTable`, `IndexRoutingTable`, `IndexShardRoutingTable` and `ShardRouting`.